firmware/stellar.c

```diff
@@ -0,0 +1,876 @@
+
+    // get full address string
+    stellar_publicAddressAsStr(addrbytes, str_fulladdr, sizeof(str_fulladdr));
+
+    // Break it into 3 lines
+    strlcpy(rows[0], str_fulladdr + 0, 17);
+    strlcpy(rows[1], str_fulladdr + 16, 21);
+    strlcpy(rows[2], str_fulladdr + 16 + 20, 21);
+
+    static const char *ret[3] = { rows[0], rows[1], rows[2] };
+    return ret;
+}
+
+size_t stellar_publicAddressAsStr(uint8_t *bytes, char *out, size_t outlen)
+{
+    // version + key bytes + checksum
+    uint8_t keylen = 1 + 32 + 2;
+    uint8_t bytes_full[keylen];
+    bytes_full[0] = 6 << 3; // 'G'
+
+    memcpy(bytes_full + 1, bytes, 32);
+
+    // Last two bytes are the checksum
+    uint16_t checksum = stellar_crc16(bytes_full, 33);
+    bytes_full[keylen-2] = checksum & 0x00ff;
+    bytes_full[keylen-1] = (checksum>>8) & 0x00ff;
+
+    base32_encode(bytes_full, keylen, out, outlen, BASE32_ALPHABET_RFC4648);
+
+    // Public key will always be 56 characters
+    return 56;
+}
+
+/*
+ * CRC16 implementation compatible with the Stellar version
+ * Ported from this implementation: http://introcs.cs.princeton.edu/java/61data/CRC16CCITT.java.html
+ * Initial value changed to 0x0000 to match Stellar
+ */
+uint16_t stellar_crc16(uint8_t *bytes, uint32_t length)
+{
+    // Calculate checksum for existing bytes
+    uint16_t crc = 0x0000;
+    uint16_t polynomial = 0x1021;
+    uint32_t i;
+    uint8_t bit;
+    uint8_t byte;
+    uint8_t bitidx;
+    uint8_t c15;
+
+    for (i=0; i < length; i++) {
+        byte = bytes[i];
+        for (bitidx=0; bitidx < 8; bitidx++) {
+            bit = ((byte >> (7 - bitidx) & 1) == 1);
+            c15 = ((crc >> 15 & 1) == 1);
+            crc <<= 1;
+            if (c15 ^ bit) crc ^= polynomial;
+        }
+    }
+
+    return crc & 0xffff;
+}
+
+/*
+ * Writes 32-byte public key to out
+ */
+void stellar_getPubkeyAtIndex(uint32_t index, uint8_t *out, size_t outlen)
+{
+    if (outlen < 32) return;
+
+    HDNode *node = stellar_deriveNode(index);
+
+    ed25519_public_key stellarPubkeyBytes;
+    ed25519_publickey(node->private_key, stellarPubkeyBytes);
+    memcpy(out, stellarPubkeyBytes, outlen);
+}
+
+/*
+ * Derives the HDNode at the given index
+ * The prefix for this is m/44'/148'/index'
+ */
+HDNode *stellar_deriveNode(uint32_t index)
+{
+    static CONFIDENTIAL HDNode node;
+    const char *curve = "ed25519";
+
+    // Derivation path for Stellar is m/44'/148'/index'
+    uint32_t address_n[3];
+    address_n[0] = 0x80000000 | 44;
+    address_n[1] = 0x80000000 | 148;
+    address_n[2] = 0x80000000 | index;
+
+    // Device not initialized, passphrase request cancelled, or unsupported curve
+    if (!storage_getRootNode(&node, curve, true)) {
+        return 0;
+    }
+    // Failed to derive private key
+    if (hdnode_private_ckd_cached(&node, address_n, 3, NULL) == 0) {
+        return 0;
+    }
+
+    // Populate public key
+    ed25519_publickey(node.private_key, node.public_key + 1);
```

Comment:
Why not use `hdnode_fill_public_key`?

Reply:
Thanks for this tip and the other suggestions!